airflow/logging/airflow_task_handler.py

```diff
@@ -0,0 +1,244 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import warnings
+
+from airflow import configuration as conf
+from airflow.configuration import AirflowConfigException
+from airflow.utils import logging as logging_utils
+from airflow.utils.file import mkdirs
+from airflow.utils.state import State
+
+
+class AirflowTaskHandler(logging.Handler):
+    """
+    AirflowTaskHandler is the python log handler that handles and reads
+    task instance logs. It creates and delegates log handling
+    to `logging.FileHandler` after receiving task instance context.
+    Upon close, it uploads the log to remote storage like S3/GCS.
+    """
+
+    def __init__(self):
+        super(AirflowTaskHandler, self).__init__()
+        # Setup dummy class variables before receiving task instance to
+        # prevent AttributeError when system shuts down.
+        self.handler = logging.NullHandler()
```

Comment:
I would prefer a generic location, in case someone forgets the set the context. Or raise an exception when the context has not been set yet and one starts logging.